File: core/src/braindustry/logic/StrOp.java
package braindustry.logic;

import arc.func.Func3;
import arc.util.Strings;

public enum StrOp {
    add("+", (String a, String b) -> a + "" + b, "a", "b"),
    add3("+ 3x", (String a, String b, String c) -> a + b + c, "a", "b", "c"),
    number("num", (String a) -> {
        if (Strings.canParseFloat(a)) return Strings.parseFloat(a);
        return null;
    }, "str"),
    str("str", (Object a) -> {
        return a + "";
    }, "obj"),
    length("length", (String a) -> {
        return a == null ? -1 : a.length();
    }, "text"),
    indexOf("index", (String a, String b, Double c) -> {
        return c == null || c <= 0 ? a.indexOf(b) : a.indexOf(b, c.intValue());
    }, "text", "str", "indexFrom"),
    lastIndexOf("lastIndex", (String a, String b, Double c) -> {
        return c == null || c <= 0 ? a.lastIndexOf(b) : a.lastIndexOf(b, c.intValue());
    }, "text", "str", "indexFrom"),
    substring("sub", (String a, double b, double c) -> {

        if (b >= 0) {
            if ((c < 0)) {
                return a.substring((int) b).intern();
            } else if (c < a.length()) {
                return a.substring((int) b, (int) c).intern();
            }
        }
        return null;
    }, "text", "from", "to"),
    starts("starts", (String a, String b) -> a.startsWith(b)),
    end("ends", (String a, String b) -> a.endsWith(b));

    public static final StrOp[] all = values();
    static final int strVal = 0, objVal = 1, numVal = 2;
    public final Func3<Object, Object, Object, Object> func;
    public final SrtOpType type;
    public final String symbol;
    public final String[] params;

    StrOp(String symbol, StrOpLambda1Str function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get((String) a);
        type = function.getType();
        this.params = params;
    }

    StrOp(String symbol, StrOpLambda1 function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get(a);
        type = function.getType();
        this.params = params;
    }

    StrOp(String symbol, StrOpLambda2 function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get(a, b);
        type = function.getType();
        this.params = params;
    }

    StrOp(String symbol, StrOpLambda2One function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get((String) a, (Double) b);
        type = function.getType();
        this.params = params;
    }

    StrOp(String symbol, StrOpLambda2All function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get((String) a, (String) b);
        type = function.getType();
        this.params = params;
    }

    StrOp(String symbol, StrOpLambda3 function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get((String) a, (Double) b, (Double) c);
        type = function.getType();
        this.params = params;
    }

    StrOp(String symbol, StrOpLambda3TwoStr function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get((String) a, (String) b, (Double) c);
        type = function.getType();
        this.params = params;
    }

    StrOp(String symbol, StrOpLambda3All function, String... params) {
        this.symbol = symbol;
        func = (a, b, c) -> function.get((String) a, (String) b, (String) c);
        type = function.getType();
        this.params = params;
    }

    @Override
    public String toString() {
        return symbol;
    }

    interface StrOpLambda3All {
        Object get(String a, String b, String c);

        default SrtOpType getType() {
            return new SrtOpType(strVal,strVal,strVal);
        }
    }

    interface StrOpLambda3TwoStr {
        Object get(String a, String b, Double c);

        default SrtOpType getType() {
            return new SrtOpType(strVal,strVal,numVal);
        }
    }

    interface StrOpLambda3 {
        Object get(String a, double b, double c);

        default SrtOpType getType() {
            return new SrtOpType(strVal,numVal,numVal);
        }
    }

    interface StrOpLambda2One {
        Object get(String a, double b);

        default SrtOpType getType() {
            return new SrtOpType(strVal,numVal,-1);
        }
    }

    interface StrOpLambda2 {
        Object get(Object a, Object b);

        default SrtOpType getType() {
            return new SrtOpType(objVal,objVal,-1);
        }
    }

    interface StrOpLambda2All {
        Object get(String a, String b);

        default SrtOpType getType() {
            return new SrtOpType(strVal,strVal,-1);
        }
    }

    interface StrOpLambda1Str {
        Object get(String a);

        default SrtOpType getType() {
            return new SrtOpType(strVal,-1,-1);
        }
    }


    interface StrOpLambda1 {
        String get(Object a);

        default SrtOpType getType() {
            return new SrtOpType(objVal,-1,-1);
        }
    }

    static class SrtOpType {
        final int first, second, third;

        SrtOpType(int first, int second, int third) {
            this.first = first;
            this.second = second;
            this.third = third;
        }
    }
}

File: core/src/braindustry/logic/StrOp.kt
@file:JvmMultifileClass

package braindustry.logic

import arc.func.Func3
import arc.math.Mathf
import mindustry.logic.LExecutor
import kotlin.math.min

object StringOperationScope {
    var a: Any? = null
    var b: Any? = null
    var c: Any? = null
    inline val strA: String
        get() = a.asLogicString();
    inline val strB: String
        get() = b.asLogicString();
    inline val strC: String
        get() = c.asLogicString();
    inline val numA: Double
        get() = strA.toDoubleOrNull() ?: 0.0;
    inline val numB: Double
        get() = strB.toDoubleOrNull() ?: 0.0;
    inline val numC: Double
        get() = strC.toDoubleOrNull() ?: 0.0;

    operator fun invoke(a: Any?, b: Any? = null, c: Any? = null): StringOperationScope {
        this.a = a;
        this.b = b;
        this.c = c;
        return StringOperationScope;
    }

    operator fun invoke(a: LExecutor.Var?, b: LExecutor.Var? = null, c: LExecutor.Var? = null): StringOperationScope {
        return this(a.unwrap(),b.unwrap(),c.unwrap())
    }

    inline fun <reified T> T.asLogicString(): String = if (this is Number) this.toString() else LExecutor.PrintI.toString(this)
    private fun LExecutor.Var?.unwrap(): Any? {
        return if (this == null) {
            null
        } else {
            (if (isobj) objval else numval)
        }
    }
}


@Suppress("EnumEntryName","unused")
enum class StrOp(@JvmField val symbol: String,@JvmField val paramsAmount: Int, block: StringOperationScope.() -> Any?, @JvmField vararg val params: String) {
    add("+", 2, { "$strA$strB" }, "a", "b"),
    add3("+ 3x", 3, { "$strA$strB$strC" }, "a", "b", "c"),
    number("num", 1, { numA }, "str"),
    str("str", 1, { strA }, "obj"),
    length("length", 1, { strA.length }, "text"),
    indexOf("index", 3, { if (numC <= 0) strA.indexOf(strB) else strA.indexOf(strB, numC.toInt()) },
        "text", "str", "indexFrom"
    ),
    lastIndexOf("lastIndex", 3, { if (numC <= 0) strA.lastIndexOf(strB) else strA.lastIndexOf(strB, numC.toInt()) },
        "text", "str", "indexFrom"
    ),
    substring("sub", 3, {
        if (numB < 0) {
            null
        } else if (numC < 0) {
            strA.substring(numB.toInt())
        } else if (numC < strA.length && numB <= numC && numC <= strA.length) {
            strA.substring(numB.toInt(), numC.toInt())
        } else {
            null
        }
    }, "text", "from", "to"),
    starts("starts", 2, { strA.startsWith(strB) }),
    end("ends", 2, { strA.endsWith(strB) }),
    chatAt("chatAt", 2, { "${strA[Mathf.clamp(numB.toInt(), 0, strA.length - 1)]}" }),
    insert("insert", 3,
        { "${strA.substring(0, min(strA.length, numB.toInt()))}$strC${strA.substring(min(strA.length-1, numB.toInt()))}" }),
    replace("replace", 3, { strA.replace(strB, strC) });

    @JvmField
    val func: Func3<LExecutor.Var, LExecutor.Var, LExecutor.Var, Any?> = when (paramsAmount) {
        1 -> Func3 { a, _, _ -> StringOperationScope(a).block() }
        2 -> Func3 { a, b, _ -> StringOperationScope(a, b).block() }
        3 -> Func3 { a, b, c -> StringOperationScope(a, b, c).block() }
        else -> {
            throw IllegalArgumentException("illegal params amount: $paramsAmount")
        }
    }


    override fun toString(): String {
        return symbol
    }


    companion object {
        @JvmField
        val all = values()
    }
}